esp-wifi/src/wifi/mod.rs

```diff
@@ -3193,247 +2713,585 @@ pub(crate) fn apply_power_saving(ps: PowerSaveMode) -> Result<(), WifiError> {
+                warn!("Failed to cleanly deinit wifi: {:?}", e);
+            }
+        }
+    }
+}
 
-            Ok((result, count))
+impl WifiController<'_> {
+    /// Attempts to take the sniffer, returns `Some(Sniffer)` if successful,
+    /// otherwise `None`.
+    #[cfg(feature = "sniffer")]
+    pub fn take_sniffer(&self) -> Option<Sniffer> {
+        if self
+            .sniffer_taken
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
+            == Ok(false)
+        {
+            Some(Sniffer::new())
+        } else {
+            None
         }
+    }
+
+    /// Set CSI configuration and register the receiving callback.
+    #[cfg(feature = "csi")]
+    pub fn set_csi(
+        &mut self,
+        mut csi: CsiConfig,
+        cb: impl FnMut(crate::wifi::wifi_csi_info_t) + Send,
+    ) -> Result<(), WifiError> {
+        csi.apply_config()?;
+        csi.set_receive_cb(cb)?;
+        csi.set_csi(true)?;
 
-        /// Async version of [`crate::wifi::WifiController`]'s `start` method
-        pub async fn start_async(&mut self) -> Result<(), WifiError> {
-            let mode = WifiMode::try_from(&self.config)?;
+        Ok(())
+    }
+
+    /// Set the wifi protocol.
+    ///
+    /// This will set the wifi protocol to the desired protocol, the default for
+    /// this is: `WIFI_PROTOCOL_11B|WIFI_PROTOCOL_11G|WIFI_PROTOCOL_11N`
+    ///
+    /// # Arguments:
+    ///
+    /// * `protocols` - The desired protocols
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// wifi_controller.set_protocol(Protocol::P802D11BGNLR.into());
+    /// ```
+    pub fn set_protocol(&mut self, protocols: EnumSet<Protocol>) -> Result<(), WifiError> {
+        let mut protocol = 0u8;
 
-            let mut events = enumset::enum_set! {};
-            if mode.is_ap() {
-                events |= WifiEvent::ApStart;
+        protocols.into_iter().for_each(|v| match v {
+            Protocol::P802D11B => protocol |= WIFI_PROTOCOL_11B as u8,
+            Protocol::P802D11BG => protocol |= WIFI_PROTOCOL_11B as u8 | WIFI_PROTOCOL_11G as u8,
+            Protocol::P802D11BGN => {
+                protocol |=
+                    WIFI_PROTOCOL_11B as u8 | WIFI_PROTOCOL_11G as u8 | WIFI_PROTOCOL_11N as u8
+            }
+            Protocol::P802D11BGNLR => {
+                protocol |= WIFI_PROTOCOL_11B as u8
+                    | WIFI_PROTOCOL_11G as u8
+                    | WIFI_PROTOCOL_11N as u8
+                    | WIFI_PROTOCOL_LR as u8
             }
-            if mode.is_sta() {
-                events |= WifiEvent::StaStart;
+            Protocol::P802D11LR => protocol |= WIFI_PROTOCOL_LR as u8,
+            Protocol::P802D11BGNAX => {
+                protocol |= WIFI_PROTOCOL_11B as u8
+                    | WIFI_PROTOCOL_11G as u8
+                    | WIFI_PROTOCOL_11N as u8
+                    | WIFI_PROTOCOL_11AX as u8
             }
+        });
```

Comment:
Meaningless nitpick of the day, but this cna be so much nicer if you collected these into a `protocol: u32` and cast it only in one place.

Reply:
I'll 2nd that and suggest using `map` and [`fold`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold) as well.